### PR TITLE
Fieldops region 2d-3d special treatment

### PR DESCRIPTION
**Pull request by d7919**
The mixed `Field3D`/`Field2D` operators in the generated fieldops are currently not getting vectorised. This PR adds in some special handling of these cases to ensure that they also get vectorised.

This should help address #928.

Related additions to mesh add routines for converting from 3d to 2d indices and back again, as well as a lookup method for converting 3d->2d that is used in `Field2D::operator[](Ind3D)`, avoiding division in each call.

### CODE REVIEW

File: include/bout/mesh.hxx [hunk 1]
```diff
@@ -597,7 +597,19 @@ class Mesh {
...
+
+  Ind3D ind2Dto3D(const Ind2D &ind2D, int jz = 0){
+    return Ind3D(ind2D.ind * LocalNz + jz);
+  }
+
+  Ind2D ind3Dto2D(const Ind3D &ind3D){
+    return Ind2D(ind3D.ind / LocalNz);
+  }
+
+  int map3Dto2D(const Ind3D &ind3D){
```
Comment (ZedThree) [inline include/bout/mesh.hxx]: Please could you add doxygen comments for these -- specifically, what's the difference between `ind3Dto2D` and `map3Dto2D`?

**Comment by ZedThree** [comment]
Looks good, just needs documentation + tests :)

File: include/bout/region.hxx [hunk 2]
```diff
@@ -204,12 +205,20 @@ public:
   }
 };
 
+/// Arithmetic operators with integers
+inline Ind3D operator+(Ind3D lhs, const Ind3D &rhs) { return lhs += rhs; }
```
[the remark this answers is not in the document]
Comment (d7919) [reply include/bout/region.hxx]: I think so, I don't think it worked without this. Also, if we allowed `SpecificInd + SpecificInd` then you could presumably do `Ind3D + Ind2D` which is not well defined.

[the remark this answers is not in the document]
Comment (ZedThree) [reply include/bout/region.hxx]: That's a good point. Might be best to remove the `SpecificInd` arithmetic operators then

**Pull request merged**